Executable fails on import setuptools

## Description of the issue

PyInstaller executable fails with `AssertionError: ...\distutils\core.pyc` when trying to `import setuptools`.

The `setuptools` override of `distutils` does not seem to be loaded, see [ensure_local_distutils][3] in setuptools source.

Same as [setuptools issue 3089][1], but, as far as I know, the issue only arises when running as a PyInstaller executable.

### Context information (for bug reports)

* Output of `pyinstaller --version`: 5.1
* Version of Python: 3.8.13
* Platform: Windows 10
* How you installed Python: conda
* Did you also try this on another platform? Does it work there? 

  did not try

* try the latest development version

  done, see below

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [ x ] start with clean installation
  * [ x ] use the latest development version
  * [ x ] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [ x ] Package your program in **--onedir mode**
  * [ x ] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [ x ] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.


### A minimal example program which shows the error

Similar to what is described in this [comment to the setuptools issue][2].

Create `repro.py` file with content:

```
import setuptools
``` 

Then on the command line:
```
conda create -n temp python=3.8
conda activate temp
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
pyinstaller --onedir --debug imports --noupx repro.py
dist\repro\repro.exe
```

Using conda here for convenience, but issue also arises using a normal python installation.

### Stacktrace / full error message


```
Traceback (most recent call last):
  File "repro.py", line 1, in <module>
    import setuptools
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "setuptools\__init__.py", line 8, in <module>
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "PyInstaller\loader\pyimod03_importers.py", line 495, in exec_module
  File "_distutils_hack\override.py", line 1, in <module>
  File "_distutils_hack\__init__.py", line 77, in do_override
  File "_distutils_hack\__init__.py", line 64, in ensure_local_distutils
AssertionError: C:\Users\me\Desktop\temp\setuptools-issue-3089\dist\repro\distutils\core.pyc
[6064] Failed to execute script 'repro' due to unhandled exception!
```

[1]: https://github.com/pypa/setuptools/issues/3089
[2]: https://github.com/pypa/setuptools/issues/3089#issuecomment-1032305835
[3]: https://github.com/pypa/setuptools/blob/main/_distutils_hack/__init__.py#L51-L65

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -14,11 +14,6 @@ on:
       - '!pyup/**'
 
 env:
-  # Prevent setuptools from replacing distutils with its vendored version, as our test suite is not
-  # ready for that. There also seem to be issues with setuptools-provided distutils and pip 22.2
-  # on python < 3.10.
-  SETUPTOOLS_USE_DISTUTILS: not-today
-
   # Colored pytest output on CI despite not having a tty
   FORCE_COLOR: 1
 
@@ -141,7 +136,6 @@ jobs:
       - run: docker build -f alpine.dockerfile -t foo .
       - run: >
           docker run
-          -e SETUPTOOLS_USE_DISTUTILS='not-today'
           foo
           pytest
           -n 3 --maxfail 3 --durations 10 tests/unit tests/functional

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -6,12 +6,6 @@ on:
       - develop
       - v4
 
-env:
-  # Prevent setuptools from replacing distutils with its vendored version. There seem to be issues
-  # between setuptools-provided distutils and pip 22.2 on python < 3.10, which results in failure
-  # of the "Test Packaging" step.
-  SETUPTOOLS_USE_DISTUTILS: not-today
-
 jobs:
   run:
     runs-on: ubuntu-latest

--- a/PyInstaller/hooks/hook-distutils.py
+++ b/PyInstaller/hooks/hook-distutils.py
@@ -8,12 +8,10 @@
 #
 # SPDX-License-Identifier: (GPL-2.0-or-later WITH Bootloader-exception)
 #-----------------------------------------------------------------------------
-"""
-`distutils`-specific post-import hook.
 
-This hook freezes the external `Makefile` and `pyconfig.h` files bundled with the active Python interpreter, which the
-`distutils.sysconfig` module parses at runtime for platform-specific metadata.
-"""
+from PyInstaller.utils.hooks import is_module_satisfies
+
+hiddenimports = []
 
 # From Python 3.6 and later ``distutils.sysconfig`` takes on the same behaviour as regular ``sysconfig`` of moving the
 # config vars to a module (see hook-sysconfig.py). It doesn't use a nice `get module name` function like ``sysconfig``
@@ -21,8 +19,15 @@ This hook freezes the external `Makefile` and `pyconfig.h` files bundled with th
 # ``_get_sysconfigdata_name()`` from regular ``sysconfig``.
 try:
     import sysconfig
-    hiddenimports = [sysconfig._get_sysconfigdata_name()]
+    hiddenimports += [sysconfig._get_sysconfigdata_name()]
 except AttributeError:
     # Either sysconfig has no attribute _get_sysconfigdata_name (i.e., the function does not exist), or this is Windows
     # and the _get_sysconfigdata_name() call failed due to missing sys.abiflags attribute.
     pass
+
+# Starting with setuptools 60.0, the vendored distutils overrides the stdlib one (which will be removed in python 3.12
+# anyway), so check if we are using that version. While the distutils override behavior can be controleld via the
+# ``SETUPTOOLS_USE_DISTUTILS`` environment variable, the latter may have a different value during the build and at the
+# runtime, and so we need to ensure that both stdlib and setuptools variant of distutils are collected.
+if is_module_satisfies("setuptools >= 60.0"):
+    hiddenimports += ['setuptools._distutils']

--- a/PyInstaller/hooks/hook-setuptools.py
+++ b/PyInstaller/hooks/hook-setuptools.py
@@ -25,7 +25,13 @@ if is_unix or is_darwin:
 # setuptools >= 39.0.0 is "vendoring" its own direct dependencies from "_vendor" to "extern". This also requires
 # 'pre_safe_import_module/hook-setuptools.extern.six.moves.py' to make the moves defined in 'setuptools._vendor.six'
 # importable under 'setuptools.extern.six'.
-hiddenimports.extend(collect_submodules('setuptools._vendor'))
+_excluded_submodules = (
+    # Prevent recursing into setuptools._vendor.pyparsing.diagram, which typically fails to be imported due to
+    # missing dependencies (railroad, pyparsing (?), jinja2) and generates a warning... As the module is usually
+    # unimportable, it is likely not to be used by setuptools.
+    'setuptools._vendor.pyparsing.diagram',
+)
+hiddenimports.extend(collect_submodules('setuptools._vendor', filter=lambda name: name not in _excluded_submodules))
 
 # As of setuptools >= 60.0, we need to collect the vendored version of distutils via hiddenimports. The corresponding
 # pyi_rth_setuptools runtime hook ensures that the _distutils_hack is installed at the program startup, which allows

--- a/PyInstaller/hooks/hook-setuptools.py
+++ b/PyInstaller/hooks/hook-setuptools.py
@@ -10,7 +10,7 @@
 #-----------------------------------------------------------------------------
 
 from PyInstaller.compat import is_darwin, is_unix
-from PyInstaller.utils.hooks import collect_submodules
+from PyInstaller.utils.hooks import collect_submodules, is_module_satisfies
 
 hiddenimports = [
     # Test case import/test_zipimport2 fails during importing pkg_resources or setuptools when module not present.
@@ -26,3 +26,10 @@ if is_unix or is_darwin:
 # 'pre_safe_import_module/hook-setuptools.extern.six.moves.py' to make the moves defined in 'setuptools._vendor.six'
 # importable under 'setuptools.extern.six'.
 hiddenimports.extend(collect_submodules('setuptools._vendor'))
+
+# As of setuptools >= 60.0, we need to collect the vendored version of distutils via hiddenimports. The corresponding
+# pyi_rth_setuptools runtime hook ensures that the _distutils_hack is installed at the program startup, which allows
+# setuptools to override the stdlib distutils with its vendored version, if necessary.
+if is_module_satisfies("setuptools >= 60.0"):
+    hiddenimports += ["_distutils_hack"]
+    hiddenimports += collect_submodules("setuptools._distutils")

--- a/PyInstaller/hooks/pre_find_module_path/hook-distutils.py
+++ b/PyInstaller/hooks/pre_find_module_path/hook-distutils.py
@@ -11,15 +11,17 @@
 """
 `distutils`-specific pre-find module path hook.
 
-When run from within a venv (virtual environment), this hook changes the `__path__` of the `distutils` package to
-that of the system-wide rather than venv-specific `distutils` package. While the former is suitable for freezing,
-the latter is intended for use _only_ from within venvs.
+When run from within a virtual environment, this hook changes the `__path__` of the `distutils` package to
+that of the system-wide rather than virtual-environment-specific `distutils` package. While the former is suitable for
+freezing, the latter is intended for use _only_ from within virtual environments.
+
+NOTE: this behavior seems to be specific to virtual environments created by (an old?) version of `virtualenv`; it is not
+applicable to virtual environments created by the `venv`.
 """
 
-import distutils
-import os
+import pathlib
 
-from PyInstaller.utils.hooks import logger
+from PyInstaller.utils.hooks import logger, get_module_file_attribute
 
 
 def pre_find_module_path(api):
@@ -28,11 +30,17 @@ def pre_find_module_path(api):
     # opcode is not a virtualenv module, so we can use it to find the stdlib. Technique taken from virtualenv's
     # "distutils" package detection at
     # https://github.com/pypa/virtualenv/blob/16.3.0/virtualenv_embedded/distutils-init.py#L5
-    import opcode
+    # As opcode is a module, stdlib path corresponds to the parent directory of its ``__file__`` attribute.
+    stdlib_path = pathlib.Path(get_module_file_attribute('opcode')).parent.resolve()
+    # As distutils is a package, we need to consider the grandparent directory of its ``__file__`` attribute.
+    distutils_path = pathlib.Path(get_module_file_attribute('distutils')).parent.parent.resolve()
 
-    system_module_path = os.path.normpath(os.path.dirname(opcode.__file__))
-    loaded_module_path = os.path.normpath(os.path.dirname(distutils.__file__))
-    if system_module_path != loaded_module_path:
-        # Find this package in its parent directory.
-        api.search_dirs = [system_module_path]
-        logger.info('distutils: retargeting to non-venv dir %r', system_module_path)
+    if distutils_path.name == 'setuptools':
+        logger.debug("distutils: provided by setuptools")
+    elif distutils_path == stdlib_path:
+        logger.debug("distutils: provided by stdlib")
+    else:
+        # Find this package in stdlib.
+        stdlib_path = str(stdlib_path)
+        logger.debug("distutils: virtualenv shim - retargeting to stdlib dir %r", stdlib_path)
+        api.search_dirs = [stdlib_path]

--- a/PyInstaller/hooks/rthooks.dat
+++ b/PyInstaller/hooks/rthooks.dat
@@ -25,5 +25,6 @@
     'win32api':   ['pyi_rth_win32api.py'],
     'win32com':   ['pyi_rth_win32comgenpy.py'],
     'multiprocessing': ['pyi_rth_multiprocessing.py'],
+    'setuptools': ['pyi_rth_setuptools.py'],
     'subprocess': ['pyi_rth_subprocess.py'],
 }

--- a/PyInstaller/hooks/rthooks/pyi_rth_setuptools.py
+++ b/PyInstaller/hooks/rthooks/pyi_rth_setuptools.py
@@ -1,0 +1,21 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2022, PyInstaller Development Team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#
+# SPDX-License-Identifier: Apache-2.0
+#-----------------------------------------------------------------------------
+
+# This runtime hook performs the equivalent of the distutils-precedence.pth from the setuptools package;
+# it registers a special meta finder that diverts import of distutils to setuptools._distutils, if
+# available.
+try:
+    import os
+    if os.environ.get("SETUPTOOLS_USE_DISTUTILS", "local") == "local":
+        import _distutils_hack
+        _distutils_hack.add_shim()
+except Exception:
+    pass

--- a/PyInstaller/loader/pyimod02_importers.py
+++ b/PyInstaller/loader/pyimod02_importers.py
@@ -379,11 +379,10 @@ class FrozenImporter:
         """
         entry_name = None  # None means - no module found in this importer.
 
-        if fullname in self.toc:
-            entry_name = fullname
-            trace("import %s # PyInstaller PYZ", fullname)
-        elif path is not None:
-            # Try to handle module.__path__ modifications by the modules themselves.
+        # Try to handle module.__path__ modifications by the modules themselves. This needs to be done first in
+        # order to support module overrides in alternative locations while we also have the original module
+        # available at non-override location.
+        if path is not None:
             # Reverse the fake __path__ we added to the package module into a dotted module name, and add the tail
             # module from fullname onto that to synthesize a new fullname.
             modname = fullname.rsplit('.')[-1]
@@ -404,6 +403,13 @@ class FrozenImporter:
                     break
             else:
                 entry_name = None
+
+        if entry_name is None:
+            # Either there was no path override, or the module was not available there. Check the fully qualified name
+            # of the module directly.
+            if fullname in self.toc:
+                entry_name = fullname
+                trace("import %s # PyInstaller PYZ", fullname)
 
         if entry_name is None:
             trace("# %s not found in PYZ", fullname)

--- a/news/7075.feature.rst
+++ b/news/7075.feature.rst
@@ -1,0 +1,2 @@
+Add support for ``setuptools``-provided ``distutils``, available since
+``setuptools >= 60.0``.

--- a/tests/functional/test_libraries.py
+++ b/tests/functional/test_libraries.py
@@ -413,3 +413,10 @@ def test_pywin32ctypes(pyi_builder, submodule):
     pyi_builder.test_source("""
         from win32ctypes.pywin32 import {0}
         """.format(submodule))
+
+
+@importorskip('setuptools')
+def test_setuptools(pyi_builder):
+    pyi_builder.test_source("""
+        import setuptools
+        """)

